### ISSUE TITLE
Sharing using the edv-client

### ISSUE THREAD

**Issue by gjgd** [digitalbazaar/edv-client#25]
Hello ! I'm trying to implement a sharing functionality using the edv-client.
More specifically I want Alice to be able to create a document that Bob is able to read.

I've considered several approaches:
- Alice writes the document to her EDV (but then how does Bob access it?)
- Alice writes the document to Bob's EDV (Bob would be able to access it, but how can Alice write to Bob's EDV?)
- Alice creates a third EDV that she has write access to and that Bob has read access to

What would be the best approach using the EdvClient's API?

**Comment by aljones15** [comment]
When inserting the EDV document you can set recipients. 
Alice will need to know an identifier for Bob (such as his didKey).


https://github.com/digitalbazaar/edv-client/blob/9e961f99158c3e29e69404595a10a5b3f2d766ae/EdvClient.js#L81-L134

As you can see here you can pass in a recipients array to `insert`.
I'm not 100% sure on the format of that array anymore, but it should follow whatever `this._createDefaultRecipients(keyAgreementKey)` does:

https://github.com/digitalbazaar/edv-client/blob/9e961f99158c3e29e69404595a10a5b3f2d766ae/EdvClient.js#L784-L792

which is an object of the format:

```js
{header: { kid: keyAgreementKey.id, alg: ''ECDH-ES+A256KW''}}
```

Basically recipients will need to a keyAgreement key for Bob.
So pass in both Alice's keyAgreementKey and Bob's keyAgreementKey using that format above and Bob should be able to read Alice's document.

**Comment by dlongley** [comment]
This can be accomplished through a combination of Authorization Capabilities ("zcaps" for short) and key agreement keys.

There are several scenarios that are supported:

1. Alice writes the document to her EDV  (but then how does Bob access it?)

Here, Alice must have access to the public key half of one of Bob's key agreement keys. Bob could publish such a key to his DID Document in the `keyAgreement` field. In that case, Alice only needs to know Bob's DID to discover it. Alternatively, Bob could communicate this public key to her in some other way.

Once Alice has Bob's key agreement key, she can then write the encrypted document to her EDV, using her own key agreement key and Bob's key agreement key as recipients. This means that once Bob gets the encrypted document, he will be able to decrypt it. However, this is insufficient for Bob to actually *access* the encrypted document.

Therefore, Alice must delegate a zcap to Bob. This zcap will contain an `invoker` field with Bob's DID, indicating that only he will be able to invoke it -- by digitally signing an invocation request that references it, using a `capabilityInvocation` key specified in his DID Document. This zcap will also specify the `invocationTarget`, i.e., the URL to the encrypted document Alice will write to her EDV -- and the URL Bob will hit with his client to invoke the zcap and retrieve the encrypted document. Alice must then give this delegated zcap to Bob so he can invoke it when he pleases.

2. Alice writes the document to Bob's EDV (Bob would be able to access it, but how can Alice write to Bob's EDV?)

Here, Bob must delegate a zcap to Alice so she can write to Bob's EDV. This follows essentially the same pattern as item 1, but reversing the parties. Bob will delegate a zcap to a URL to an encrypted document on his EDV to Alice, where the zcap's `invoker` field expresses Alice's DID. Bob must still communicate the public key half of his key agreement key to Alice, either through his DID Document or some other means.

Once Alice has the delegated zcap in hand, she can encrypt the document using her and Bob as recipients (or, more accurately, the public key half of their key agreement keys as recipients) and then send it to Bob's EDV by hitting the `invocationTarget` in the zcap using her client and invoking the zcap by signing it using the private key half of a key pair, where the public key half (or "verification method") is listed in her DID Document under `capabilityInvocation`.

3. Alice creates a third EDV that she has write access to and that Bob has read access to

This is essentially a repeat of the first option, but with a new EDV.

---

Note, there are some variants to the above that may also be considered. One involves the combination of EDV and WebKMS technologies. Instead of Alice or Bob using their own key agreement keys, either Alice or Bob could generate a new key agreement key in a WebKMS system and then delegate access to *it* via a zcap. This would allow a single key agreement key to be used as a recipient on documents that Alice and Bob want to share access to. They both will be able to encrypt/decrypt using the same key, but neither of them will have direct access to the key material for that key. Instead, they will use zcaps to access the key -- as well as using zcaps to access the encrypted documents they create.

This approach can make it easier to share with more parties -- and in a way that doesn't require the re-encryption of documents when you need to add additional recipients. Instead, new zcaps are delegated to the additional recipients without having to modify the encrypted documents. Similarly, these zcaps to use the shared key agreement key can be revoked to remove recipients.

@gjgd, we have code that implements these scenarios, but we need to distill it to simple, runnable examples with Alice and Bob. @mattcollier, @aljones15 -- We need to put together some code examples to show how this is done.

**Comment by gjgd** [comment]
Thanks @aljones15 and @dlongley for the detailed explanation this is super helpful!

I was able to add Bob as a recipient to Alice's encrypted document in her EDV using the example you provided, but I'm not exactly sure how to generate the zcap to enable Bob to access Alice's EDV.

I'm guessing it has to do with this method: https://github.com/digitalbazaar/edv-client/blob/9e961f99158c3e29e69404595a10a5b3f2d766ae/EdvClient.js#L558

Do you have any examples of how to use it?

**Comment by aljones15** [comment]
@gjgd I'm expanding test coverage today to cover it. Basically you make a zCap, sign it, and pass it to that function for each recipient. Glad to see you made it and sorry my examples are lagging so behind.

**Comment by gjgd** [comment]
FYI, largely thanks to the example in this PR:  https://github.com/digitalbazaar/edv-client/pull/26 we wrote a test to show how Alice can share a document in her EDV to Bob:

https://github.com/transmute-industries/transmute-edv/blob/9a3964b733abbe6c50693e8e648a35f783a2efd4/packages/edv-app/src/store/edv/share.spec.js

I'm closing this, thanks @aljones15 and @dlongley !

**Comment by aljones15** [comment]
> FYI, largely thanks to the example in this PR: #26 we wrote a test to show how Alice can share a document in her EDV to Bob:
> 
> https://github.com/transmute-industries/transmute-edv/blob/9a3964b733abbe6c50693e8e648a35f783a2efd4/packages/edv-app/src/store/edv/share.spec.js
> 
> I'm closing this, thanks @aljones15 and @dlongley !

you're welcome. You linked to a private repo. Anyways, all the sharing stuff is changing really soon so the tests will be updated when that happens.

**Comment by gjgd** [comment]
My bad, here is a gist showing the test: https://gist.github.com/gjgd/11cac221f51a8a29c41af9ea482a18fe

**Comment by aljones15** [comment]
> My bad, here is a gist showing the test: https://gist.github.com/gjgd/11cac221f51a8a29c41af9ea482a18fe

that's cool. an integration test using `jest`. thanks for sharing.
